demos/mqtt/mqtt_demo_connection_sharing/mqtt_demo_connection_sharing.c

```diff
@@ -0,0 +1,1517 @@
+    }
+    else
+    {
+        /* This example publishes with QOS1. */
+        outgoingPublishPackets[ publishIndex ].pubInfo.qos = MQTTQoS1;
+        outgoingPublishPackets[ publishIndex ].pubInfo.pTopicName = pTopic;
+        outgoingPublishPackets[ publishIndex ].pubInfo.topicNameLength = topicLength;
+        outgoingPublishPackets[ publishIndex ].pubInfo.pPayload = pMessage;
+        outgoingPublishPackets[ publishIndex ].pubInfo.payloadLength = strlen( pMessage );
+
+        /* Get a new packet id. */
+        outgoingPublishPackets[ publishIndex ].packetId = MQTT_GetPacketId( pMqttContext );
+
+        /* Send PUBLISH packet. */
+        mqttStatus = MQTT_Publish( pMqttContext,
+                                   &outgoingPublishPackets[ publishIndex ].pubInfo,
+                                   outgoingPublishPackets[ publishIndex ].packetId );
+
+        if( mqttStatus != MQTTSuccess )
+        {
+            LogError( ( "Failed to send PUBLISH packet to broker with error = %u.",
+                        mqttStatus ) );
+            cleanupOutgoingPublishAt( publishIndex );
+            returnStatus = EXIT_FAILURE;
+        }
+        else
+        {
+            LogInfo( ( "PUBLISH sent for topic %.*s to broker with packet ID %u.\n\n",
+                       topicLength,
+                       pTopic,
+                       outgoingPublishPackets[ publishIndex ].packetId ) );
+
+            /* Calling MQTT_ProcessLoop to process incoming publish echo, since
+             * application subscribed to the same topic the broker will send
+             * publish message back to the application. This function also
+             * sends ping request to broker if MQTT_KEEP_ALIVE_INTERVAL_SECONDS
+             * has expired since the last MQTT packet sent and receive
+             * ping responses. */
+            mqttStatus = MQTT_ProcessLoop( pMqttContext, MQTT_PROCESS_LOOP_TIMEOUT_MS );
+
+            if( mqttStatus != MQTTSuccess )
+            {
+                LogWarn( ( "MQTT_ProcessLoop failed: Error = %s.",
+                           MQTT_Status_strerror( mqttStatus ) ) );
+            }
+        }
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static int subscribePublishLoop( NetworkContext_t * pNetworkContext )
+{
+    int returnStatus = EXIT_SUCCESS;
+    bool mqttSessionEstablished = false, sessionPresent;
+    MQTTContext_t mqttContext;
+    MQTTSubscribeInfo_t pSubscriptionList[ 3 ];
+
+    /* Establish MQTT session on top of TCP+TLS connection. */
+    LogInfo( ( "Creating an MQTT connection to %.*s.",
+               BROKER_ENDPOINT_LENGTH,
+               BROKER_ENDPOINT ) );
+
+    /* Sends an MQTT Connect packet using the established TLS session,
+     * then waits for connection acknowledgment (CONNACK) packet. */
+    returnStatus = establishMqttSession( &mqttContext, pNetworkContext, false, &sessionPresent );
```

Comment:
I understand if this was a requirement for this demo, but if it was only a design choice, does this demo *need* non-clean sessions? I feel like connection sharing and resending of publishes are separate features, and if we already show it in the TLS and mutual auth demos, then we don't need to show it again here. It would reduce the amount of code that's duplicated in multiple places.

Reply:
That's a good suggestion in terms of reducing the number of lines of code in this demo. I have removed the session restore feature (of resending publish messages) from this demo.